### PR TITLE
[docs]: Improve GenerateDocsAPIData for submodule generation

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

When GenerateDocsAPIData is provided an array of entrypoints it will filter the generated types to only the included files. The existing logic only matched on the filename meaning it could have false positive matches.

This is important for `expo-router` where we will have API docs for `expo-router` and `expo-router/ui`, but we need to ensure only files from the correct submodule are provided. Both submodules have files with the name filename which is causing false positives in the filtering.

# How

I've refactored the logic to compare the `sourceFileName` against the allow list of files to include.

I've also renamed some variables as I found their names too similar to other variables.

# Test Plan

Running `et generate-docs-api-data` changed no files. Running the docs things looked normal.

I'm not very familiar with the docs, so I could appreciate if someone double checked this for me.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by Simek — COMMENTED** [review]
The `sourceFileName` field will be stipped from the final results because we do not use it yet, since it's a new addition since `0.26` release:
* https://github.com/expo/expo/blob/7a5a9e5ebc37dc3a9f7abf009be38155611c5ffe/tools/src/commands/GenerateDocsAPIData.ts#L194

We handle type links wia custom mechanism, since this information was not available for a long time, and we had to develop different approach.

So to summarize, we can ship this change for the future, when we would try to switch our linking to build in TypeDoc mechanism, but at this moment it will not change the docs output in any way, beside the minified/debug output, which is not used.

**Comment by marklawlor** [comment]
This PR is pulled from this PR https://github.com/expo/expo/pull/31767. I'm having difficulty creating docs for a submodule `expo-router/ui`
